interfacer/src/browsh/vim_mode.go

```diff
@@ -0,0 +1,594 @@
+	case "scrollUp":
+		doScroll(0, -1)
+	case "scrollDown":
+		doScroll(0, 1)
+	case "scrollLeft":
+		doScroll(-1, 0)
+	case "scrollRight":
+		doScroll(1, 0)
+	case "editURL":
+		urlBarFocusToggle()
+	case "firstTab":
+		switchToTab(tabsOrder[0])
+	case "lastTab":
+		switchToTab(tabsOrder[len(tabsOrder)-1])
+	case "scrollHalfPageDown":
+		_, height := screen.Size()
+		doScroll(0, (height-uiHeight)/2)
+	case "scrollHalfPageUp":
+		_, height := screen.Size()
+		doScroll(0, -((height - uiHeight) / 2))
+	case "historyBack":
+		sendMessageToWebExtension("/tab_command,/history_back")
+	case "historyForward":
+		sendMessageToWebExtension("/tab_command,/history_forward")
+	case "reload":
+		sendMessageToWebExtension("/tab_command,/reload")
+	case "prevTab":
+		prevTab()
+	case "nextTab":
+		nextTab()
+	case "previouslyVisitedTab":
+		previouslyVisitedTab()
+	case "newTab":
+		createNewEmptyTab()
+	case "removeTab":
+		removeTab(CurrentTab.ID)
+	case "restoreTab":
+		restoreTab()
+	case "duplicateTab":
+		duplicateTab(CurrentTab.ID)
+	case "moveTabLeft":
+		moveTabLeft(CurrentTab.ID)
+	case "moveTabRight":
+		moveTabRight(CurrentTab.ID)
+	case "copyURL":
+		clipboard.WriteAll(CurrentTab.URI)
+	case "openClipboardURL":
+		URI, _ := clipboard.ReadAll()
+		sendMessageToWebExtension("/tab_command,/url," + URI)
+	case "openClipboardURLInNewTab":
+		URI, _ := clipboard.ReadAll()
+		sendMessageToWebExtension("/new_tab," + URI)
+	case "focusFirstTextInput":
+		sendMessageToWebExtension("/tab_command,/focus_first_text_input")
+	case "followLinkLabeledNext":
+		sendMessageToWebExtension("/tab_command,/follow_link_labeled_next")
+	case "followLinkLabeledPrevious":
+		sendMessageToWebExtension("/tab_command,/follow_link_labeled_previous")
+	case "viewHelp":
+		sendMessageToWebExtension("/new_tab,https://www.brow.sh/docs/keybindings/")
+	case "openLinkInCurrentTab":
+		currentVimMode = linkMode
+		sendMessageToWebExtension("/tab_command,/get_clickable_hints")
+		eraseLinkHints()
+	case "openLinkInNewTab":
+		currentVimMode = linkModeNewTab
+		sendMessageToWebExtension("/tab_command,/get_link_hints")
+		eraseLinkHints()
+	case "copyLinkURL":
+		currentVimMode = linkModeCopy
+		sendMessageToWebExtension("/tab_command,/get_link_hints")
+		eraseLinkHints()
+	case "findText":
+		fallthrough
+	case "findNext":
+		sendMessageToWebExtension("/tab_command,/find_next," + findText)
+	case "findPrevious":
+		sendMessageToWebExtension("/tab_command,/find_previous," + findText)
+	case "makeMark":
+		currentVimMode = makeMarkMode
+	case "gotoMark":
+		currentVimMode = gotoMarkMode
+	case "insertMode":
+		currentVimMode = insertMode
+	case "findMode":
+		currentVimMode = findMode
+	case "normalMode":
+		currentVimMode = normalMode
+	case "visualMode":
+		currentVimMode = visualMode
+	case "caretMode":
+		currentVimMode = caretMode
+		width, height := screen.Size()
+		caretPos.X, caretPos.Y = width/2, height/2
+	}
+}
+
+func searchScreenForText(text string) []Coordinate {
```

Comment:
Should this also scroll to the currently found text hint?

Reply:
This was my first implementation of link clicking by typing in and searching for text on the screen. It works, too. But it's not enabled by default, so my plan was to allow for this with some alternative key combination or making it a configurable link selection mode replacing the normal mode. It's possible to click all text with this, not just links, which might enable a few additional use cases.

Reply:
No this shouldn't scroll. This is intended to search for text in the currently visible screen only, hence the name searchScreenForText. This makes it very fast/efficient as no communication with the backend is required in this process. However in the future this could be combined with the find (/) feature. I.e. when the typed text isn't found on the screen a search for it could be initiated over the whole page (using the same code as the find (/) feature), it would scroll there if scrolling was fixed for the find feature (we talked about this before), and then the searchScreenForText feature could be used a second time in order to click on the text.

Reply:
Ok, no problem. I just added the word `visible` to the function name.

So anyway, my point was, that when I use this feature the screen doesn't scroll when focussing matches outside the screen. Is that intended?

Reply:
@tombh 
I'm not sure we're talking about the same thing here. The function you reference wasn't enabled in the current code and got replaced by the vimium link hinting based navigation. Unless you changed some code to make this functionality active again, you can't possibly have tried it.

The find feature (activated by the / key like in vim) is something completely different and is implemented in the webextension code. searchScreenForText and / have nothing to do with each other at the moment, but they could be combined in the future (what I was talking about in the last comment).

The find feature (/ key) can't scroll at the moment because of the scrolling not synchronizing issue I wrote you about.

To be honest I dislike your choice of adding visible to the name of searchScreenForText, because I can't imagine a logical scenario where someone would search an invisible screen. You can only search what's on the screen, and that's visible, or otherwise it wouldn't be on the screen. The word screen refers to the terminal here, not the screen of the browser backend.

Reply:
Ok, understood.

So what about providing a setting for changing the navigation behaviour? So others can try it out, I mean give it a _go_ :D Otherwise how do you feel about removing the unused code? We could leave a comment referencing the commit hash so future devs can find it again?

Regarding `visible`; is not the default behaviour of search in a browser to search the whole document, including the invisible areas?

And where shall we make a note about the broken scrolling? Is that something I could fix now for this PR?

Reply:
@tombh 
We should provide a setting for it! It could also be used in parallel to link hinting. Ie. by simply binding it to a different key.

Again the searchScreenForText function used by findAndHighlightTextOnScreen, which was written for the purpose of link navigation by text based clicking. This isn't intended as the find text feature, which indeed should scroll and search the whole page. The find text feature (like vim / and also bound to /) is implemented correctly and the window.find method used in the webextension does scrolling by itself. Unfortunately browsh somehow overwrites that scrolling, therefore it doesn't really. This could and should be fixed. You wrote to me about this, saying that we also have to solve this for pages that want to scroll to a #hash position, pages that scroll to comments and the like.

Reply:
Ok, this is bringing back memories. I can fix the scrolling. The only difficulty I think will be the potential for positive feedback loops. I'll need to place a scroll listener that will trigger for both internally automated scrolling (eg; anchor tags) and Browsh initiated scrolling. Browsh scrolls its own copy of the page and then tells the main web page to scroll, the main web page then receives a scroll trigger and tells Browsh of the new scroll position, which Browsh should honour. So either there needs to be a way to distinguish Browsh-intiated scrolling from internal scrolling, or the scrolling values have to be precisely identical.

Reply:
@tombh 
Makes sense. Fixing this will be a huge step forward. Thank you for looking into it.